### ISSUE TITLE
Distro versions and Python versions

### ISSUE THREAD

**Issue by bennuttall** [piwheels/packages#210]
piwheels builds wheels for Python versions in versions of Debian.

This issue states the current status of each Debian version in piwheels, and will be updated as things change.

| Debian version | Python version | Status |
| - | - | - |
| Jessie | 3.4 | :checkered_flag: end-of-life |
| Stretch | 3.5 | :heavy_check_mark: building |
| Buster | 3.7 | :heavy_check_mark: building |
| Bullseye | 3.9 | :clock1: coming soon |

**Comment by bennuttall** [comment]
Bullseye wheels have been working through the backlog for the last couple of weeks - we've cleared the backlog and they're now up-to-date and will continue to build from now on. :partying_face:

**Comment by atharvai** [comment]
Any roadmap for Python 3.10? I run HomeAssistant which uses numpy and compiling numpy on Raspberry Pi 2B is slow but also doesn't finish. a pre-built wheel for armv7l would be great to use

**Comment by bennuttall** [comment]
We only build for versions of Python in Debian releases. The next Debian release is in the summer, so we'll start building for that once it's released, with whatever version of Python it comes with, presumably 3.10: https://packages.debian.org/bookworm/python3

**Comment by okhiroyuki** [comment]
@bennuttall 
Debian 12 is scheduled for release on June 10. With the release of Debian 12, will piwheel start building Python 3.11?

> Release date
> ============
> 
> We plan to release on 2023-06-10.
> 
> If you want to celebrate it, please consider attending a Debian
> release party, or hosting your own! See
> https://wiki.debian.org/ReleasePartyBookworm for more information.
> (Note: the release process typically takes the whole day and the
> release isn't done until the early hours of Sunday UTC.)
> https://lists.debian.org/debian-devel-announce/2023/04/msg00007.html?ref=news.itsfoss.com

**Comment by bennuttall** [comment]
Of course

**Comment by dequis** [comment]
Hi! Approximately when can we expect bookworm packages to be available?

Thank you for piwheels <3

**Comment by bennuttall** [comment]
Waiting on some work on the backend to support multi-digit Python versions :laughing: https://github.com/piwheels/piwheels/issues/288 hopefully not too much longer before we can get that done

**Comment by holta** [comment]
News about Raspberry Pi OS 13 "Trixie" has emerged in recent days, just FYI:

https://forums.raspberrypi.com/viewtopic.php?t=389477

**Comment by MichaIng** [comment]
How long is building wheels for Bullseye/Python 3.9 planned? At first I thought it was stopped already, since I see this for a number of them:

<img width="530" height="96" alt="Image" src="https://github.com/user-attachments/assets/bdb09f63-b26e-4d9e-8421-2ae70cad5f19" />

But then I see a very recent `cryptography` Bullseye/Python 3.9 build, where instead one for Python 3.11 has this "unexpected error":

<img width="630" height="147" alt="Image" src="https://github.com/user-attachments/assets/2711e273-3d34-42e0-98a4-4a05fa394bc6" />

The dash points to an empty log file, so looks like the build failed on a very early level.

**Comment by bennuttall** [comment]
We will still build Python 3.9 until Bullseye EOL next summer.

The table works well for pure python builds and platform wheels. We can fail on one ABI and succeed on the next, or succeed on all ABIs.

The unexpected error / missing icon happens when we build an `abi3` wheel, or could be something else.

For cryptography 46.0.1, we had 2 fails and a successful build, but there are no files, indicating something went wrong. I'll try and rebuild that one.

**Comment by MichaIng** [comment]
Okay, I'll open issues for the missing (Bullseye) wheels I am aware of then.

`abi3` wheels seem to be actually a good thing: you could compile with Python 3.9 and that one wheel could be used with Python 3.11 and Python 3.13 as well ... in theory. But dynamically linked shared libraries can cause issues, of course.

I read in some other issue that tag versions are currently not sorted correctly (`cp39` > `cp311`) and hence wheels may be compiled on the wrong Debian version or so?

What I saw in some of your failing build logs, and what we face as well when trying to compile those wheels on Bullseye, is:
```
Collecting MarkupSafe>=0.9.2
  Downloading markupsafe-3.0.3.tar.gz (80 kB)
ERROR: Exception:
Traceback (most recent call last):
  File "/usr/share/python-wheels/resolvelib-0.5.4-py2.py3-none-any.whl/resolvelib/resolvers.py", line 171, in _merge_into_criterion
    crit = self.state.criteria[name]
KeyError: 'markupsafe'
```
The Bullseye/Python 3.9 `resolvelib` seems to have issues with some newer wheels, while they do not declare incompatibility with Python 3.9, as in theory it works with just newer `resolvelib`, I guess. But installing [`resolvelib`](https://pypi.org/project/resolvelib/) with `pip` into the use `venv` or so does not work either, it keeps using the instance from `/usr/share/python-wheels`. But I better open a new issue for this to collect some info.
